File: examples/python/django/core/autonoma_config.py
# =============================================================================
# Autonoma SDK Configuration (Factory-driven)
# =============================================================================
# The SDK is factory-driven: every model the dashboard can create has a
# registered factory whose `input_model` (Pydantic) drives both validation
# and the discover schema. There is no SQL introspection, no SQL fallback,
# and no executor — your factories use whatever Django ORM access your app
# already has.

import os

from pydantic import BaseModel, ConfigDict

from autonoma.types import HandlerConfig
from autonoma.factory import define_factory
from autonoma_django import create_django_handler

from core.repositories.organization import OrganizationRepository
from core.repositories.user import UserRepository


# ---------------------------------------------------------------------------
# 1. Initialize repositories
# ---------------------------------------------------------------------------
organization_repo = OrganizationRepository()
user_repo = UserRepository()


# ---------------------------------------------------------------------------
# 2. Declare the factory input/ref models
# ---------------------------------------------------------------------------
class OrganizationInput(BaseModel):
    model_config = ConfigDict(extra="ignore")
    name: str


class OrganizationRef(BaseModel):
    id: str
    name: str


class UserInput(BaseModel):
    model_config = ConfigDict(extra="ignore")
    email: str
    name: str
    organization_id: str


# ---------------------------------------------------------------------------
# 3. Build the handler config
# ---------------------------------------------------------------------------
config = HandlerConfig(
    scope_field="organization_id",
    shared_secret=os.environ.get("AUTONOMA_SHARED_SECRET", "my-shared-secret"),
    signing_secret=os.environ.get("AUTONOMA_SIGNING_SECRET", "my-signing-secret"),

    # Required: the endpoint returns 404 unless this is True. The SDK never
    # inspects PYTHON_ENV/ENV — tie it to your own condition to keep it off in
    # prod, e.g. allow_production=settings.DEBUG.
    allow_production=True,

    factories={
        "Organization": define_factory(
            create=lambda data, ctx: organization_repo.create({"name": data.name}),
            teardown=lambda record, ctx: organization_repo.delete(record.id),
            input_model=OrganizationInput,
            ref_model=OrganizationRef,
        ),

        "User": define_factory(
            create=lambda data, ctx: user_repo.create(
                {
                    "email": data.email,
                    "name": data.name,
                    "organization_id": data.organization_id,
                },
            ),
            input_model=UserInput,
        ),
    },

    auth=lambda user, context: {
        "headers": {"Authorization": "Bearer test-token"}
    },
)


# ---------------------------------------------------------------------------
# 4. Create the Django view handler
# ---------------------------------------------------------------------------
# Returns a Django view function decorated with @csrf_exempt and @require_POST.
handler = create_django_handler(config)

File: examples/java/spring-boot/src/main/java/com/example/autonoma/AutonomaConfig.java
// =============================================================================
// Autonoma SDK — Spring Boot Example (Factory-driven)
// =============================================================================
// The SDK is factory-driven: every model the dashboard can create has a
// registered factory whose inputClass (a Java record) drives both validation
// and the discover schema. There is no SQL introspection, no JDBC executor,
// and no SQL fallback — your factories call whatever services your app has.

package com.example.autonoma;

import ai.autonoma.spring.AutonomaController;
import ai.autonoma.sdk.FactoryUtil;
import ai.autonoma.sdk.types.AuthResult;
import ai.autonoma.sdk.types.HandlerConfig;
import com.example.autonoma.repositories.OrganizationRepository;
import com.example.autonoma.repositories.UserRepository;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import javax.sql.DataSource;
import java.util.Map;

@Configuration
public class AutonomaConfig {

    // Input records — drive both validation and discover schema
    public record OrganizationInput(String name) {}
    public record UserInput(String email, String name, String organizationId) {}

    @Bean
    public AutonomaController autonomaController(DataSource dataSource) {
        OrganizationRepository organizationRepo = new OrganizationRepository(dataSource);
        UserRepository userRepo = new UserRepository(dataSource);

        HandlerConfig config = new HandlerConfig(
            // The column that scopes all models to a tenant — used to isolate test data
            "organization_id",
            // Shared with Autonoma — verifies incoming requests via HMAC-SHA256
            System.getenv("AUTONOMA_SHARED_SECRET") != null ? System.getenv("AUTONOMA_SHARED_SECRET") : "my-shared-secret",
            // Private to your server — signs the refs token so teardown only deletes what was created
            System.getenv("AUTONOMA_SIGNING_SECRET") != null ? System.getenv("AUTONOMA_SIGNING_SECRET") : "my-signing-secret",
            // Called after `up` — returns credentials so Autonoma can make authenticated requests
            (user, context) -> AuthResult.ofHeaders(
                Map.of("Authorization", "Bearer test-token")
            )
        );

        // Required: the endpoint returns 404 unless this is true. The SDK never
        // inspects JAVA_ENV/SPRING_PROFILES_ACTIVE — tie it to your own condition
        // to keep it off in prod, e.g. !"production".equals(System.getenv("JAVA_ENV")).
        config.setAllowProduction(true);

        // Every model the dashboard can create needs a factory.
        // The factory's inputClass drives both validation and discover.
        config.setFactories(Map.of(
            "Organization", FactoryUtil.defineFactory(
                OrganizationInput.class,
                (data, ctx) -> organizationRepo.create(data),
                (record, ctx) -> organizationRepo.delete((String) record.get("id"))
            ),
            // data is automatically deserialized into UserInput
            "User", FactoryUtil.defineFactory(
                UserInput.class,
                (data, ctx) -> userRepo.create(data)
            )
        ));

        return new AutonomaController(config);
    }
}
